Add some unsafe magic for minimal optimizations, add .idea to .gitignore and fix some typos

Kleinmarb opened the pull request:
Thanks for the amazing contribution to open source url parsing!

I am using this lib right now and thought that I should look into it a bit,
I ended up doing minor changes.

I understand if the unsafe code gets rejected, but I even left some safety comments.

anonrig commented on src/lib.rs:
```diff
@@ -75,16 +76,18 @@ pub enum HostType {
+            return Self::Domain;
         }
+
+        // Safety: Prior to transmuting we checked if value is bigger than 2
```
Can you add the mapping (number to enum) as a comment to here? Let's not also remove the knowledge we have since it might not be obvious at the first glance.

anonrig commented on src/lib.rs:
```diff
@@ -98,16 +101,13 @@ pub enum SchemeType {
         }
+
+        // Safety: Prior to transmuting we checked if value is bigger than 6
+        // and returned the default in that case.
```
Can you add the mapping as a comment here as well?

anonrig reviewed (approved):
Nice job! Can you double check if we have tests covering the changed lines? Just in case.

anonrig commented:
Also can you fix clippy errors?

anonrig commented on src/lib.rs:
```diff
@@ -75,16 +76,18 @@ pub enum HostType {
+
+        // Safety: Prior to transmuting we checked if value is bigger than 2
+        // and returned the default in that case.
+        unsafe { std::mem::transmute(value) }
```
It appears that it is extremely unsafe. Ref: https://x.com/tenellous/status/1819506534675894296

Is there any benchmark result you could share with us, that makes this change/risk worthy to land?

Kleinmarb commented:
I added tests and comments that illustrate which number maps to which enum variant.

Regarding the clippy errors, I don't get any.

Now let's talk about performance:
If I'm honest there will be a performance difference but it will be negligible.
But the code I wrote is 100% safe, you can trust me on that one since I added a`#[repr(u32)]` on the enum which allows for safe transmution between `value as u32` and the enum as long as `value` is not bigger than the number of variants of the enum -1, but we check for that.

If you still think its risky, just deny the commit, it's alright for me.

obi1kenobi commented on src/lib.rs:
```diff
@@ -75,16 +76,24 @@ pub enum HostType {
+            return Self::Domain;
         }
+
+        // Safety: Prior to transmuting we checked if value is bigger than 2
+        // and returned the default in that case.
+        //
+        // Mapping:
+        // - 0 => Domain
+        // - 1 => IPV4
+        // - 2 => IPV6
+        // - any other number maps to NotSpecial
+        unsafe { std::mem::transmute(value as u32) }
```
```suggestion
        match value {
            0 => Self::Domain,
            1 => Self::IPV4,
            2 => Self::IPV6,
            _ => Self::Domain,
        }
```
If you replace the transmute with the same `match` block as before, the generated code (and therefore the performance) will be identical: https://x.com/tenellous/status/1819528511201464821

I'm all for performance improvement, but here there's only `unsafe` risk without any benefit.

obi1kenobi reviewed (commented):
I'm not the maintainer so this isn't my call, but for my taste this isn't worth it when the same exact benefit can be obtained without any `unsafe`.

anonrig reviewed (changes requested):
Rejecting due to https://x.com/tenellous/status/1819528511201464821?s=46

Kleinmarb commented:
Oh ok I should have done the assembly check myself. Thanks for pointing out that it produces the same exact instructions.
